### ISSUE TITLE
Cannot Backward Twice Through Distribution

### ISSUE THREAD

**Issue by purboo** [pytorch/pytorch#11192]
## Issue description

Cannot backward twice through distribution (see example below).

## Code example

It is ok to backward through two different loss:

```
t = torch.tensor(1.0, requires_grad=True)
l1 = 2 * t
l2 = 3 * t

l1.backward() # ok
l2.backward() # ok
```

However, it is not OK to backward through two different loss through distribution:

```
p = torch.tensor(0.2, requires_grad=True)
pi = Bernoulli(p)

l1 = 2 * pi.log_prob(torch.tensor(1.0)) 
l2 = 3 * pi.log_prob(torch.tensor(2.0))
l1.backward() # ok
l2.backward() # Trying to backward through the graph a second time, but the buffers have already been freed. Specify retain_graph=True when calling backward the first time.
```

What is the reason behind this? 

## System Info

- PyTorch or Caffe2: Pytorch
- How you installed PyTorch (conda, pip, source): pip
- OS: Arch Linux
- PyTorch version: 0.4.1.post2
- Python version: 3.7.0
- CUDA/cuDNN version: 9.0
- GPU models and configuration: Maxwell Titan X
- GCC version (if compiling from source): 8.1.0

**Comment by purboo** [comment]
Thanks a lot for the explanation.

**Comment by purboo** [comment]
@SsnL I found something strange. Very appreciate if you could help me clear my doubts:

### Case 3:

```
t = torch.tensor(1.0, requires_grad=True)
k = 2 * t
l1 = 2 * k
l2 = 3 * k

l1.backward() # ok
l2.backward() # ok
```

In this case, 

```
t ----> k ----> l1
             \
                -----> l2
```

However, after the graph `t ---> k ----> l1` is freed, the graph `t ---> k ---> l2` is still complete? 

### Case 4

The error occurs when changing `k = 2*t` to `k = t*t`:

```
t = torch.tensor(1.0, requires_grad=True)
k = t * t
l1 = 2 * k
l2 = 3 * k

l1.backward() # ok
l2.backward() # Trying to backward through the graph a second time, but the buffers have already been freed. Specify retain_graph=True when calling backward the first time.
```

Is case 4 different from case 3? Thank you.

**Comment by SsnL** [comment]
@purboo Not all tensors from a graph is needed. Some times you only need gradients to compute a backward, e.g., when multiplied with a scalar C, the gradient for input is just C times the gradient for output. :) So even if the tensor containing the output value is freed, you can still backprop for scalar multiplication. But this is not the case for f(x)=x^2.

**Comment by purboo** [comment]
@SsnL Many thanks for the clarification!
